### PR TITLE
fix(menu): multiple close events for a single close

### PR DESCRIPTION
**Pull request by ppham27**
Don't emit a closed event when another event will be emitted. Previously, if one clicked on a menu item, one would get two events: `undefined` and `click` in that order. One would see similar behavior for `keydown` or clicking the backdrop. Unit tests were updated to prevent a regression.

### CODE REVIEW

File: src/lib/menu/menu-directive.ts [hunk 1]
```diff
@@ -139,7 +139,7 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
   }
 
   /** Event emitted when the menu is closed. */
-  @Output() close = new EventEmitter<void | 'click' | 'keydown'>();
+  @Output() close = new EventEmitter<void | 'click' | 'destroy' | 'keydown'>();
```
Comment (crisbeto) [inline src/lib/menu/menu-directive.ts]: The idea with this one was to emit the reason only if it's user-generated. Given that there is no special behavior for `destroy`, can you revert it back to `void | 'click' | 'keydown'`?

File: src/lib/menu/menu-trigger.ts [hunk 2]
```diff
@@ -216,23 +216,34 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
...
+  /** Focuses the menu trigger. */
+  focus() {
+    this._element.nativeElement.focus();
+  }
+
+  /**
+   * Gives the option of setting emitEvent to false to avoid emitting an event
+   * when an event will already be emitted.
+   */
+  private _closeMenu(emitEvent: boolean) {
```
Comment (crisbeto) [inline src/lib/menu/menu-trigger.ts]: Rather than having the `emitEvent` parameter, you can remove the `this.menu.close.emit` call from here and move it only to `closeMenu`. Also change the doc string to something along the lines of "Closes the menu and does the necessary cleanup".

File: src/lib/menu/menu.spec.ts [hunk 3]
```diff
@@ -485,6 +485,7 @@ describe('MdMenu', () => {
       fixture = TestBed.createComponent(SimpleMenu);
       fixture.detectChanges();
       fixture.componentInstance.trigger.openMenu();
+      fixture.componentInstance.closeCallback.calls.reset();
```
Comment (crisbeto) [inline src/lib/menu/menu.spec.ts]: Why is this necessary? The callbacks should be reset after each test, because the component is re-compiled.

File: src/lib/menu/menu.spec.ts [hunk 4]
```diff
@@ -493,26 +494,40 @@ describe('MdMenu', () => {
...
 
       backdrop.click();
       fixture.detectChanges();
 
-      expect(fixture.componentInstance.closeCallback).toHaveBeenCalled();
+      expect(fixture.componentInstance.closeCallback).toHaveBeenCalledWith(undefined);
+      expect(fixture.componentInstance.closeCallback).toHaveBeenCalledTimes(1);
+    });
+
+    it('should emit an event when escaped', () => {
```
Comment (crisbeto) [inline src/lib/menu/menu.spec.ts]: "should emit an event when pressing ESCAPE".

File: src/lib/menu/menu.spec.ts [hunk 5]
```diff
@@ -989,11 +1004,18 @@ describe('MdMenu', () => {
 
       expect(menus.length).toBe(3, 'Expected three open menus');
 
+      instance.rootCloseCallback.calls.reset();
```
Comment (crisbeto) [inline src/lib/menu/menu.spec.ts]: These reset calls shouldn't be necessary.

Comment (ppham27) [reply src/lib/menu/menu.spec.ts]: They were necessary when because Observable.of(null) was triggering a close event. Replaced with Observable.empty()

File: src/lib/menu/menu.spec.ts (hunk 4 above)
Comment (ppham27) [reply src/lib/menu/menu.spec.ts]: Done.

File: src/lib/menu/menu.spec.ts (hunk 3 above)
Comment (ppham27) [reply src/lib/menu/menu.spec.ts]: They were necessary when because Observable.of(null) was triggering a close event. Replaced with Observable.empty()

File: src/lib/menu/menu-trigger.ts (hunk 2 above)
Comment (ppham27) [reply src/lib/menu/menu-trigger.ts]: Done. Should I only emit the event when the menu is already open, though?

File: src/lib/menu/menu-directive.ts (hunk 1 above)
Comment (ppham27) [reply src/lib/menu/menu-directive.ts]: Done. I was also wondering if it's a bug that destroy will emit a close event when the menu is already closed?

Comment (crisbeto) [reply src/lib/menu/menu-directive.ts]: It's there to notify any child menus to close.

File: src/lib/menu/menu.spec.ts (hunk 3 above)
Comment (crisbeto) [reply src/lib/menu/menu.spec.ts]: `Observable.of(null)` will emit `null` immediately. `Observable.of()` won't.

File: src/lib/menu/menu-directive.ts (hunk 1 above)
Comment (ppham27) [reply src/lib/menu/menu-directive.ts]: But child menus won't be open unless the menu itself is open? Anyway, I'm more than happy to leave it as it is, as it doesn't really affect any use cases that I can think of.

Comment (crisbeto) [reply src/lib/menu/menu-directive.ts]: In most cases the children won't be open without a parent, but it's possible for the parent to be destroyed while the children are still open.

File: src/lib/menu/menu.spec.ts (hunk 3 above)
Comment (ppham27) [reply src/lib/menu/menu.spec.ts]: Are you saying that I should change it to Observable.of() instead of Observable.empty()? To me, Observable.empty() is more readable and makes the intent more clear.

But I don't mind change it to Observable.of() since the behavior would be the same as Observable.empty().

Comment (crisbeto) [reply src/lib/menu/menu.spec.ts]: Observable.empty() is definitely valid in this case. I just want to avoid bringing in an extra symbol from RxJS.

File: src/lib/menu/menu-trigger.ts [hunk 6]
```diff
@@ -216,23 +218,29 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
...
+
+  /** Focuses the menu trigger. */
+  focus() {
+    this._element.nativeElement.focus();
+  }
+
+  /**
+   * Closes the menu and does the necessary cleanup.
+   */
+  private _closeMenu() {
```
Comment (crisbeto) [inline src/lib/menu/menu-trigger.ts]: One last nitpick: can you rename this to `_destroyMenu` to avoid confusion with `closeMenu`?

File: src/lib/menu/menu.spec.ts (hunk 3 above)
Comment (ppham27) [reply src/lib/menu/menu.spec.ts]: Changed to Observable.of()

File: src/lib/menu/menu-trigger.ts (hunk 6 above)
Comment (ppham27) [reply src/lib/menu/menu-trigger.ts]: Done.
